extensions/pkg/webhook/utils_test.go

```diff
@@ -0,0 +1,682 @@
+					},
+					{
+						Name:  "envVar2",
+						Value: "value2",
+					},
+				}))
+			})
+
+			It("should do nothing to the existing EnvVar if it's identical", func() {
+				identicalEnvVar := envVars[0]
+				result := webhook.EnsureEnvVarWithName(envVars, identicalEnvVar)
+				Expect(result).To(Equal(envVars))
+			})
+		})
+
+		Describe("EnsureUnitOption", func() {
+			var unitOptions []*unit.UnitOption
+
+			BeforeEach(func() {
+				unitOptions = []*unit.UnitOption{
+					{
+						Section: "Unit",
+						Name:    "Description",
+						Value:   "Test Unit 1",
+					},
+					{
+						Section: "Service",
+						Name:    "ExecStart",
+						Value:   "/usr/bin/test1",
+					},
+				}
+			})
+
+			It("should add a new UnitOption if not present", func() {
+				newOption := &unit.UnitOption{
+					Section: "Install",
+					Name:    "WantedBy",
+					Value:   "multi-user.target",
+				}
+				result := webhook.EnsureUnitOption(unitOptions, newOption)
+				Expect(result).To(Equal(append(unitOptions, newOption)))
+			})
+
+			It("should not replace the existing UnitOption if it's not identical but add it", func() {
+				existingOption := &unit.UnitOption{
+					Section: "Unit",
+					Name:    "Description",
+					Value:   "Test Unit 2",
+				}
+				result := webhook.EnsureUnitOption(unitOptions, existingOption)
+				Expect(result).To(Equal(append(unitOptions, existingOption)))
+			})
+
+			It("should do nothing to the existing UnitOption if it's identical", func() {
+				identicalOption := unitOptions[0]
+				result := webhook.EnsureUnitOption(unitOptions, identicalOption)
+				Expect(result).To(Equal(unitOptions))
+			})
+		})
+
+		Describe("EnsureAnnotationOrLabel", func() {
+			var annotations map[string]string
+
+			BeforeEach(func() {
+				annotations = map[string]string{
+					"annotation1": "value1",
+					"annotation2": "value2",
+				}
+			})
+
+			It("should ensure the specified annotation or label exists", func() {
+				result := webhook.EnsureAnnotationOrLabel(annotations, "annotation3", "value3")
+				Expect(len(result)).To(Equal(3))
+				Expect(result["annotation3"]).To(Equal("value3"))
+			})
+
+			It("should overwrite the value of an existing annotation or label", func() {
+				result := webhook.EnsureAnnotationOrLabel(annotations, "annotation1", "newvalue1")
+				Expect(len(result)).To(Equal(2))
+				Expect(result["annotation1"]).To(Equal("newvalue1"))
+			})
+
+			It("should create a new map if the input map is nil", func() {
+				var nilMap map[string]string
+				result := webhook.EnsureAnnotationOrLabel(nilMap, "annotation1", "value1")
+				Expect(len(result)).To(Equal(1))
+				Expect(result["annotation1"]).To(Equal("value1"))
+			})
+		})
+
+		Describe("EnsureStringWithPrefix", func() {
+			var flags []string
+
+			BeforeEach(func() {
+				flags = []string{
+					"--prefix1=key1=value1,key2=value2,key3=value3",
+					"--prefix2=key4=value4,key5=value5,key6=value6",
+					"--prefix1=key7=value7,key8=value8,key9=value9",
```

Comment:
A more real-world example (which is the motivation for #8296) would be:
```suggestion
				flags = []string{
					"--prefix1-flag1=value1",
					"--flag2=value2",
					"--prefix1-flag3=value3",
				}
```

Idk why you put many key-value pairs as flag values. In the common case, flag supports a single value. You could make your flags everywhere single-value ones to make the tests easier to read.

Reply:
Thanks for the better real-world test. The problem is that for `EnsureNoStringWithPrefixContains` and `EnsureStringWithPrefixContains` without a breaking change these scenarios can't be covered. We can:
1. Cover these cases by introducing breaking change to the implementation of `EnsureNoStringWithPrefixContains` and `EnsureStringWithPrefixContains`
2. Leave those cases and cover only 
```
--prefix1=value1,value3
--prefix2=value2
--prefix1=value1,value4
```

Reply:
What is exactly the problem with these funcs? 

They seem to behave as expected according to my testing:

```diff
diff --git a/extensions/pkg/webhook/utils_test.go b/extensions/pkg/webhook/utils_test.go
index 37d8f0b80..59d72234b 100644
--- a/extensions/pkg/webhook/utils_test.go
+++ b/extensions/pkg/webhook/utils_test.go
@@ -501,24 +501,24 @@ var _ = Describe("Utils", func() {
 
 			BeforeEach(func() {
 				flags = []string{
-					"--flag1=key1=value1,key2=value2,key3=value3",
-					"--flag2=key4=value4,key5=value5,key6=value6",
-					"--flag1=key7=value7,key8=value8,key9=value9",
+					"--prefix1-flag1=value1",
+					"--flag2=value2",
+					"--prefix1-flag3=value3",
 				}
 			})
 
 			It("should ensure the specified key-value pair is in all flags with a given prefix", func() {
-				result := webhook.EnsureStringWithPrefixContains(flags, "--flag1=", "key3=value3", ",")
+				result := webhook.EnsureStringWithPrefixContains(flags, "--prefix1", "value4", ",")
 				Expect(result).To(Equal([]string{
-					"--flag1=key1=value1,key2=value2,key3=value3",
-					"--flag2=key4=value4,key5=value5,key6=value6",
-					"--flag1=key7=value7,key8=value8,key9=value9,key3=value3",
+					"--prefix1-flag1=value1,value4",
+					"--flag2=value2",
+					"--prefix1-flag3=value3,value4",
 				}))
 			})
 
 			It("should add a new flag with the specified key-value pair if no flag with the given prefix exists", func() {
-				result := webhook.EnsureStringWithPrefixContains(flags, "--flag3=", "key10=value10", ",")
-				Expect(result).To(Equal(append(flags, "--flag3=key10=value10")))
+				result := webhook.EnsureStringWithPrefixContains(flags, "--flag3=", "value4", ",")
+				Expect(result).To(Equal(append(flags, "--flag3=value4")))
 			})
 		})
 	})
@@ -663,18 +663,18 @@ var _ = Describe("Utils", func() {
 
 			BeforeEach(func() {
 				flags = []string{
-					"--prefix1=key1=value1,key2=value2,key3=value3",
-					"--prefix2=key4=value4,key5=value5,key6=value6",
-					"--prefix1=key7=value7,key8=value8,key9=value9",
+					"--prefix1-flag1=value1,value2",
+					"--flag2=value2",
+					"--prefix1-flag3=value3",
 				}
 			})
 
 			It("should delete the specified key-value pair from all flags with a given prefix", func() {
-				result := webhook.EnsureNoStringWithPrefixContains(flags, "--prefix1=", "key2=value2", ",")
+				result := webhook.EnsureNoStringWithPrefixContains(flags, "--prefix1", "value2", ",")
 				Expect(result).To(Equal([]string{
-					"--prefix1=key1=value1,key3=value3",
-					"--prefix2=key4=value4,key5=value5,key6=value6",
-					"--prefix1=key7=value7,key8=value8,key9=value9",
+					"--prefix1-flag1=value1",
+					"--flag2=value2",
+					"--prefix1-flag3=value3",
 				}))
 			})
 		})
```

Am I missing something?

Reply:
Currently those incorrect tests pass:
``` golang
			It("should ensure the specified key-value pair is in all flags with a given prefix", func() {
				result := webhook.EnsureStringWithPrefixContains(flags, "--prefix1", "value1", ",")
				Expect(result).To(Equal([]string{
					"--prefix1-flag1=value1,value1",
					"--flag2=value2",
					"--prefix1-flag3=value3,value1",
				}))
			})

			It("should ensure the specified key-value pair is in all flags with a given prefix", func() {
				result := webhook.EnsureStringWithPrefixContains(flags, "--prefix1", "-flag1=value1", ",")
				Expect(result).To(Equal([]string{
					"--prefix1-flag1=value1",
					"--flag2=value2",
					"--prefix1-flag3=value3,-flag1=value1",
				}))
			})
```
Because of this logic on line 150
```golang
		values := strings.Split(strings.TrimPrefix(items[i], prefix), sep)
```

Reply:
I see. Then let's add tests for the flag prefix only to the funcs that support it. As shared above, such flags with common prefixes are the motivation for the issue this PR is fixing, hence it would be nice to cover these scenarios with unit tests. You init the flags in BeforeEach for every Describe, hence we could have different flags (with and without prefixes) for the different Describes.

Reply:
We had a call with @Kostov6 and we agreed to:
1. Add a doc string to `EnsureStringWithPrefixContains` that it does not work well in all cases when a flag prefix (not the whole flag) is passed
2. Adapt the tests for `EnsureNoStringWithPrefix` to include a test case for deletion by flag prefix
3. Randomize the used flags as test data - as shared above, right now we only have flags that have key-value pairs as flag value. Instead, let's have flags with single value, multiple values and key-value pairs.